catalyst/callbacks/metric.py

```diff
@@ -173,46 +182,91 @@ def __init__(self, **kwargs):
         """
         super().__init__(**kwargs)
 
-        self.input = defaultdict(lambda: [])
-        self.output = defaultdict(lambda: [])
+        self.input: Optional[Dict[str, Iterable[Any]]] = None
+        self.output: Optional[Dict[str, Iterable[Any]]] = None
+        self._storage_size: int = 0
+        self._cur_idx: int = 0
 
     def on_loader_start(self, runner: "IRunner"):
         """Reinitialises internal storage."""
-        self.input = defaultdict(lambda: [])
-        self.output = defaultdict(lambda: [])
+        self.input = defaultdict(lambda: np.ndarray(0))
+        self.output = defaultdict(lambda: np.ndarray(0))
+
+    def on_batch_start(self, runner: "IRunner"):
+        """Allocate memory for data accumulation"""
+        output = self._get_output(runner.output, self.output_key)
+        input = self._get_input(runner.input, self.input_key)
+
+        dataset = runner.loaders[runner.loader_key].dataset
+        self._storage_size = len(dataset)
+
+        is_first_batch = not (self.input and self.output)
+        if not is_first_batch:
+            return
+
+        for data, storage in zip((input, output), (self.input, self.output)):
+            if isinstance(data, dict):
+                for key, value in data.items():
+                    value_shape = value.shape
+                    field_shape = (
+                        (self._storage_size, value_shape[1])
+                        if len(value_shape) > 1
+                        else (self._storage_size,)
+                    )
+                    storage[key] = np.empty(shape=field_shape)
+            else:
+                storage["_data"] = np.empty(shape=(self._storage_size,))
+        self._cur_idx = 0
 
     def on_batch_end(self, runner: "IRunner") -> None:
         """Stores new input/output for the metric computation."""
         output = self._get_output(runner.output, self.output_key)
         input = self._get_input(runner.input, self.input_key)
 
+        value_shape = 0
         for data, storage in zip((input, output), (self.input, self.output)):
             if isinstance(data, dict):
                 for key, value in data.items():
-                    storage[key].append(value.detach().cpu().numpy())
+                    value_shape = value.shape[0]
+                    storage[key][
+                        self._cur_idx : self._cur_idx + value_shape
+                    ] = (value.detach().cpu().numpy())
             else:
-                storage["_data"].append(data.detach().cpu().numpy())
+                value_shape = len(data)
+                storage["_data"][
+                    self._cur_idx : self._cur_idx + value_shape
+                ] = (data.detach().cpu().numpy())
+        self._cur_idx += value_shape
+
+    def _check_completeness(self) -> None:
+        """Check if callback allocated all the data from loader"""
+        assert self._cur_idx == self._storage_size, (
```

Comment:
let's use warning here, please
assert could cause an error during the checking experiment run